go/src/lib/lomcommon/helper.go

```diff
@@ -464,3 +583,354 @@ func (p *oneShotTimer_t) runOneShotTimer() {
     }
 }
 
+/****************************End Oneshor timer************************************************/
+
+/*************************************************************************************************/
+/* Goroutine Tracker																			 */
+/*************************************************************************************************/
+
+// GoroutineTracker is a helper for tracking goroutines. It can be used to
+// a) Track all goroutines
+// b) wait on specific goroutine to finish
+// c) Get status of all goroutines
+
+type GoroutineStatus int
+
+const (
+	GoroutineStatusRunning GoroutineStatus = iota
+	GoroutineStatusFinished
+)
+
+type Goroutine struct {
+	status    GoroutineStatus
+	done      chan struct{}
+	startTime time.Time
+	endTime   time.Time
+	args      interface{}
+}
+
+type GoroutineInfo struct {
+	Name      string
+	Status    GoroutineStatus
+	StartTime time.Time
+	EndTime   time.Time
+	Duration  time.Duration
+	Args      interface{}
+}
+
+type GoroutineTracker struct {
+	mlock      sync.Mutex
+	goroutines map[string]*Goroutine
+	waitGroup  sync.WaitGroup
+}
+
+// Create a new GoroutineTracker and return GoroutineTracker instance
+// Returns true if the goroutine is runing, false otherwise.
+//
+// Parameters:
+// - None:
+//
+// Returns:
+// - pointer to a new GoroutineTracker instance:
+func NewGoroutineTracker() *GoroutineTracker {
+	return &GoroutineTracker{
+		mlock:      sync.Mutex{},
+		goroutines: make(map[string]*Goroutine),
+		waitGroup:  sync.WaitGroup{},
+	}
+}
+
+// Start a goroutine with given name. If a goroutine with same name already exists, then it panics
+// Returns true if the goroutine is runing, false otherwise.
+//
+// Parameters:
+// - name: the name of the goroutine
+// - fn  : Function to be called
+// - args - Arguments to be passed to the function
+//
+// Returns:
+// - None:
+func (grt *GoroutineTracker) Start(name string, fn interface{}, args ...interface{}) {
+	grt.mlock.Lock()
+	defer grt.mlock.Unlock()
+
+	if _, ok := grt.goroutines[name]; ok {
+		// Goroutine with same name already exists
+		panic(fmt.Sprintf("Cannot start goroutine. Name %q already exists", name))
+	}
+
+	g := &Goroutine{status: GoroutineStatusRunning, done: make(chan struct{}), startTime: time.Now(), args: args}
+	grt.goroutines[name] = g
+	grt.waitGroup.Add(1)
+
+	// Start the goroutine
+	go func() {
+		defer func() {
+			grt.mlock.Lock()
+			defer grt.mlock.Unlock()
+
+			g.status = GoroutineStatusFinished
+			g.endTime = time.Now()
+			close(g.done)
+			grt.waitGroup.Done()
+		}()
+
+		// Get the reflect.Value of the function
+		f := reflect.ValueOf(fn)
+
+		// If the function is not a valid type, panic with an error message
+		if f.Kind() != reflect.Func {
+			panic("Invalid function type")
```

Comment:
Generally this must be code bug. So I decided to panic. Starting with same must be avoided.  For example request() is called with some anomaly name and if code again calls it again? Best way is to  use getUUID() always with goroutine name. so that it will provide unique name always.